src/ActionList/List.tsx

```diff
@@ -0,0 +1,69 @@
+import {Group, GroupProps} from './Group'
+import {Item, ItemProps} from './Item'
+import React from 'react'
+import {Divider} from './Divider'
+import {Header, HeaderProps} from './Header'
+import styled from 'styled-components'
+import {get} from '../constants'
+
+type Flatten<T extends unknown> = T extends (infer U)[] ? U : never
+
+interface UngroupedListProps {
+  items: ItemProps[]
+  renderItem?: (props: ItemProps) => JSX.Element
+}
+function isUngroupedListProps(props: ListProps): props is UngroupedListProps {
+  return typeof props === 'object' && props !== null && !('groupMetadata' in props)
+}
+
+interface GroupedListProps extends UngroupedListProps {
+  groupMetadata: (GroupProps & {groupId: number; header?: HeaderProps})[]
+  items: (ItemProps & {groupId: number})[]
+}
+function isGroupedListProps(props: ListProps): props is GroupedListProps {
+  return typeof props === 'object' && props !== null && 'groupMetadata' in props
+}
+
+export type ListProps = UngroupedListProps | GroupedListProps
+
+const StyledList = styled.div`
+  font-size: ${get('fontSizes.1')};
+`
+
+export function List({renderItem = Item, ...props}: ListProps): JSX.Element {
+  const toJSX = (itemProps: ItemProps) =>
+    'renderItem' in itemProps ? itemProps.renderItem(itemProps) : renderItem(itemProps)
+
+  const groups = (() => {
+    if (isUngroupedListProps(props)) {
+      return [{items: props.items?.map(toJSX)}]
+    } else if (isGroupedListProps(props)) {
+      const groupMap = props.groupMetadata.reduce(
+        (groups, groupMetadata) => groups.set(groupMetadata.groupId, groupMetadata),
+        new Map<number, Omit<Flatten<GroupedListProps['groupMetadata']>, 'groupId'> & {items?: JSX.Element[]}>()
+      )
+      props.items.forEach(itemProps => {
+        const group = groupMap.get(itemProps.groupId)
+        groupMap.set(itemProps.groupId, {
+          ...group,
+          items: [...(group?.items ?? []), toJSX({renderItem: group?.renderItem ?? renderItem, ...itemProps})]
+        })
+      })
```

Comment:
I would like to have a lint rule that forbids `forEach`, preferring `for..of`. The latter is more performant (no cost of instantiating and calling a function over and over), versatile (works with all iterables, not just arrays; also works with `await`, and imo easier to read.

```suggestion
      for (const itemProps of props.items) {
        const group = groupMap.get(itemProps.groupId)
        groupMap.set(itemProps.groupId, {
          ...group,
          items: [...(group?.items ?? []), toJSX({renderItem: group?.renderItem ?? renderItem, ...itemProps})]
        })
      }
```

Reply:
Done! https://github.com/primer/components/pull/1078/commits/7f321c6a80859ccd016e86b2e15d32b21b849033